### ISSUE TITLE
Celery busy-waits on the result under gevent/eventlet

### ISSUE THREAD

**Issue by stefanmh** [celery/celery#4030]
## Steps to reproduce

PoC:

```python
import gevent.monkey
gevent.monkey.patch_all()

import traceback
import greenlet
import time
import sys
import os

from celery import Celery

app = Celery('busylooping', broker='redis://', backend='redis://')

num_switches = 0

def trace_tracker(event, args):
	global num_switches
	if event in ('switch', 'throw'):
		# Uncomment to get the busy waiting traceback
		# origin, target = args
		# print 'Switching from\n%s\n\n\nto\n\n\n%s' % (
		# 	''.join(traceback.format_list(traceback.extract_stack(origin.gr_frame))),
		# 	''.join(traceback.format_list(traceback.extract_stack(target.gr_frame)))
		# )
		num_switches += 1


@app.task
def add(x, y):
	time.sleep(3.0)
	return x + y

def main1():
	app.worker_main()

def main2():
	greenlet.settrace(trace_tracker)
	add.s(3, 3).apply_async().get()
	print 'context switches:', num_switches

def main():
	if os.getenv('X') == '1':
		main1()
	else:
		main2()

if __name__ == '__main__':
	main()


```

In the first terminal, run `X=1 python celery_busylooping.py`.
In the second terminal, run `python celery_busylooping.py`

## Expected behavior

Wait on an event instead of spinning.

## Actual behavior

```
➜  ~/tmp python celery_busylooping.py
context switches: 1148048
```

## Cause

See https://github.com/celery/celery/blob/master/celery/backends/async.py#L97

Traceback:
```
Traceback (most recent call last):
  File "celery_busylooping.py", line 47, in <module>
    main()
  File "celery_busylooping.py", line 44, in main
    main2()
  File "celery_busylooping.py", line 37, in main2
    print add.s(3, 3).apply_async().get()
  File "***/local/lib/python2.7/site-packages/celery/result.py", line 189, in get
    on_message=on_message,
  File "***/local/lib/python2.7/site-packages/celery/backends/async.py", line 189, in wait_for_pending
    for _ in self._wait_for_pending(result, **kwargs):
  File "***/local/lib/python2.7/site-packages/celery/backends/async.py", line 258, in _wait_for_pending
    sleep(0)
  File "/usr/lib/python2.7/dist-packages/gevent/hub.py", line 192, in sleep
    waiter.get()
  File "/usr/lib/python2.7/dist-packages/gevent/hub.py", line 875, in get
    return self.hub.switch()
  File "/usr/lib/python2.7/dist-packages/gevent/hub.py", line 606, in switch
    return greenlet.switch(self)
```

**Comment by auvipy** [comment]
does this still exists in master will all new dependencies?

**Comment by alexpirine** [comment]
I have some data that shows that switching to gevent makes the CPU usage slowly increase over time, and redis RAM as well. I believe this issue is not closed yet.

**Comment by thedrow** [comment]
> I have some data that shows that switching to gevent makes the CPU usage slowly increase over time, and redis RAM as well. I believe this issue is not solved yet.
> 
> I guess this is because I call some tasks without calling .get(), .forget() or passing ingore_result=True.

Can you please open a new issue. This one has become really stale.
